### DISCUSSION TITLE
Usage with redux toolkit

### DISCUSSION THREAD

**Discussion by evertbouw** [redux-observable/redux-observable#792] [Ideas]
I might turn this into a docs page. Let me know if this is clear enough.

Basic gist for typescript is that you don't create a union of all your action types because they will just be `string`. Use the `.match` on an action instead to have the payload typed in your epics.

For JavaScript just delete everything with a squiggly red line

```ts
import {
  AnyAction,
  combineReducers,
  configureStore,
  createSlice,
  PayloadAction,
  getDefaultMiddleware
} from "@reduxjs/toolkit";
import { createEpicMiddleware, Epic } from "redux-observable";
import { filter, map, delay } from "rxjs/operators";

export const counter = createSlice({
  name: "counter",
  initialState: 0,
  reducers: {
    increment: (state, action: PayloadAction<number>) => state + action.payload,
    decrement: (state, action: PayloadAction<number>) => state - action.payload
  }
});

const reducer = combineReducers({
  counter: counter.reducer
});

export type MyState = ReturnType<typeof reducer>;

export type MyEpic = Epic<AnyAction, AnyAction, MyState>;

const countEpic: MyEpic = action$ =>
  action$.pipe(
    filter(counter.actions.increment.match),
    delay(500),
    map(action => counter.actions.decrement(action.payload))
  );

const epicMiddleware = createEpicMiddleware<AnyAction, AnyAction, MyState>();

export const store = configureStore({
  reducer,
  middleware: [
    ...getDefaultMiddleware({
      thunk: false // or true if you want to use thunks
    }),
    epicMiddleware
  ]
});

epicMiddleware.run(countEpic);
 ```

**Comment by jayphelps** [comment]
Hmm I'm not yet familiar with the createSlice stuff, just seen it around Twitter.

> Basic gist for typescript is that you **don't create a union of all your action types because they will just be string**. Use the .match on an action instead to have the payload typed in your epics.

I don't follow this statement, can you clarify? Creating a union of _string literal types_ (not `string`) and then narrowing that union has been the way for quite a while, but sounds like the recommendation maybe changed recently in redux? And links?

i.e.
```ts
type Actions = "foo" | "bar";
```

Would _not_ be a string type, it would be a union of the two string literals.

**Comment by evertbouw** [comment]
Because with createSlice you give a name and object key, and it merges this together for the action type, it's not possible to turn this into a string literal. If you do want to use string literals you'd have to use createAction, but this adds more boilerplate and doesn't gain you anything 

```ts
const increment = createAction<number, "counter/increment">("counter/increment");
const decrement = createAction<number, "counter/decrement">("counter/decrement");

export const counter = createReducer(0, builder => builder
  .addCase(increment, (state, action) => state + action.payload)
  .addCase(decrement, (state, action) => state - action.payload)
);

const reducer = combineReducers({
  counter
});

export type MyState = ReturnType<typeof reducer>;
type IncrementAction = ReturnType<typeof increment>;
type DecrementAction = ReturnType<typeof decrement>;
export type MyActions = IncrementAction | DecrementAction
export type MyEpic<Out extends MyActions = MyActions> = Epic<MyActions, Out, MyState>;

const countEpic: MyEpic<DecrementAction> = action$ =>
  action$.pipe(
    // generic can be removed in the next version of redux-observable
    // can also use `filter(increment.match)`
    ofType<MyActions, IncrementAction>(increment.type),
    delay(500),
    map(action => decrement(action.payload))
  );

const epicMiddleware = createEpicMiddleware<MyActions, MyActions, MyState>();
```

**Comment by jayphelps** [comment]
Ah gotcha. Sounds like I need to learn more about createSlice and the other new stuff.

Btw, IMO ideally you want the output union of your Epic to be narrowed enough to provide you with confirmation your epic emits what you think it emits, rather than AnyAction. I'm not sure if it's possible with createSlice alone (sounds like the same problem).

**Comment by evertbouw** [comment]
You might be able to catch some errors by doing
```ts
export type MyEpic<Out extends AnyAction = AnyAction> = Epic<AnyAction, Out, MyState>;

const countEpic: MyEpic<ReturnType<typeof counter.actions.decrement>> = action$ =>
```
but since in my example increment and decrement both have a signature of `{ type: string, payload: number` it'd allow both actions to be returned. It would at least check that the payload is the right type.


Updated my second example a bit and it's using `createReducer` instead of `createSlice`. It's the same number of LoCs and type safe so maybe that's the better approach

**Comment by asherccohen** [comment]
Hi guys, I'll jump into the discussion as I've been working on a few migrations from redux to the new redux-toolkit syntax which included extensive redux-observable configuration.

Historically union types were the way to go for redux configuration, but that has changed.

Redux-toolkit is now suggesting a new approach to type actions and provides utilities that help removing a lot of code.

Action types are now hidden as an implementation detail of createAction or createSlice (the second being the most suggested way of creating a reducer slice nowadays).

```
const appSlice = createSlice({
  name: "app",
  initialState,
  reducers: {
    // Actions that are used in epics with no reducer can be written like this:
    test: (state) => state,
    tested: (state) => state,
  },
});
```

Redux toolkit provides a utility fn for its actions (a match method) that narrows down types in filters.

```
export const test = (
  action$: Observable<AnyAction>,
  state$: StateObservable<RootState>
) => action$.pipe(filter(appActions.test.match), map(appActions.tested));
```

These changes have a consequence on types and as Lenz very well describes here [https://phryneas.de/redux-typescript-no-discriminating-union](url) type guards are a more powerful approach to union types (on literal types) that avoid "lying" to the compiler.

My point here is simply that it would be nice to have a section "Usage with Typescript" that describes this new approach (usage of the AnyAction type and matcher), to align with the same section in redux-toolkit.

Now, while we're discussing about this I also noticed something that wasn't super clear to me and required a few hours of digging.

The createEpicMiddleware has a type definition that gives a default param of void to state:

`export declare function createEpicMiddleware<T extends Action, O extends T = T, S = void, D = any>(options?: Options<D>): EpicMiddleware<T, O, S, D>;
`

To my understanding this is because the most common approach is to not use state in epics (I guess to leave state management to reducers), as explained here: [.https://github.com/redux-observable/redux-observable/issues/564](url)

Unfortunately some of us do, and automatically place a RootState as third arg of the EpicMiddleware type like so:

```
export const epicMiddleware: EpicMiddleware<
  AnyAction,
  AnyAction,
  RootState
> = createEpicMiddleware();
```

At this point if we're writing an epic that doesn't use state, we'll spend a few hrs debugging why epicMiddleware.run(epics) complains about state not matching. 

This might have just caught me unprepared but would be nice to have a little note in the troubleshooting saying:
"Please be careful, if epicMiddleware.run complains about state, check your epics and if none of them uses state, don't type it".

**Comment by evertbouw** [comment]
```js
    // Actions that are used in epics with no reducer can be written like this:
    test: (state) => state,
```
Hmmm I'd prefer to use `createAction` for this

**Comment by nikasepiskveradze** [comment]
Hi guys, do you have any idea how to get rid of this line `epicMiddleware.run(countEpic);` when using redux toolkit. I think it's annoying when you manually import all your async actions one place. Can we achieve redux thunk behavior, when you can just define your async actions and thats all.

**Comment by whazor** [comment]
**Thunks in Redux-observable (experimental)**
In refactoring of a code base I am moving away from redux-toolkit thunks and I made redux-observable thunks. With `createThunkAction` to create three actions at once, `ReturnThunkType` for combining three action types in one, and `thunk` function to simplify the for me most common epic. Any suggestions or feedback are welcome. I am particular keen in looking if the thunk could become more composable without having to de-thunk 😄.

**edit:** I created a NPM package with these helpers: https://github.com/Whazor/redux-observable-thunk

**edit 2 (8 Oct 2021):** Just released 2.0.0-beta.0 with support for rxjs 7 and redux-observable 2. Brought a very nice feature `filterActions`, where `filterActions(actionA, actionB)` will correctly filter redux toolkit actions (with the types working! thanks @dykam for help with typescript generics). Also a **breaking change**, thunk now only accepts toolkit actions without `.match`. Finally, there is also support for redux toolkit async thunks. Check the readme for more information.

**Example**
Below we have the create environment thunk as an example of how the helpers are used.

Code to create all the actions, the `createEnvironmentActions` is used next for the action types. Exported are the actions named `createEnvironment`, `createEnvironmentFulfilled`, `createEnvironmentFailure`.
```typescript
export const {actions: createEnvironmentActions, actions: {
  request: createEnvironment, 
  fulfilled: createEnvironmentFulfilled, 
  failure: createEnvironmentFailure
}} = createThunkActions("environments/createEnvironment", {
    request: withPayload<EnvironmentDef>(),
    fulfilled: withPayload<{ id: number }>(),
    failure: withPayload<{ error: string }>()
  });
```

As mentioned, this would combine the three `ReturnType` of each action into one action (so you don't accidentally miss one).
```typescript
export type EnvironmentActions = 
   ReturnThunkType<typeof createEnvironmentActions>
   // | ...
   ;
```

The thunk: it matches the action, acquires the latest state (not used in this example), provides dependencies, and performs a `mergeMap` on the function you provide (as seen below).
Note, in new version of my package, `.match` is removed.
```typescript
export const createEnvironmentThunk = thunk(createEnvironment.match,
  (action, _state, {request}) => {
    return request<{id: number}>(`/api/environments`, {
      method: "post",
      body: action.payload
    }).pipe(
      map(response => createEnvironmentFulfilled(response)),
      catchError((error: string) => of(createEnvironmentFailure({ error })))
    )
  });
```
<details>
  <summary>Source of all helpers</summary>
  
```typescript
import { Action } from "redux";
import { RootAction, RootEpicType } from "../state/root";
import { EpicDependencies, RootState } from "../state/store";
import { Observable } from "rxjs";
import { filter, mergeMap, withLatestFrom } from "rxjs/operators";
import { createAction, PrepareAction } from "@reduxjs/toolkit";

export function withPayload<T>() {
  return (t: T) => ({ payload: t });
}

export function thunk<A extends RootAction>(
  match: (action: Action<any>) => action is A, 
  result: (action: A, state: RootState, dependencies: EpicDependencies) => Observable<RootAction>): RootEpicType {
  return (action$, state$, dependencies) => action$.pipe(
      filter(match),
      withLatestFrom(state$),
      mergeMap(([action, state]) => result(action, state, dependencies))
  )
}

export function createThunkActions<
  PARequest extends PrepareAction<any>,
  PASuccess extends PrepareAction<any>,
  PAFailure extends PrepareAction<any>
>(name: string,
  options: {
    request: PARequest,
    fulfilled: PASuccess,
    failure: PAFailure
  }) {
    return {
      actions: {
        request: createAction(name, options.request),
        fulfilled: createAction(name + "Fulfilled", options.fulfilled),
        failure: createAction(name + "Failure", options.failure)
      }
    }
  }

export type ReturnThunkType<A> = A extends ReturnType<typeof createThunkAction>["actions"] ? 
  ReturnType<A["request"]> | 
  ReturnType<A["fulfilled"]> | 
  ReturnType<A["failure"]> : never
```
</details>

**Comment by nixxquality** [comment]
Thanks for the discussion above, it was very helpful!

----

Just putting it here in case anyone else also reads this thread after googling.

I had a workflow of react-redux, redux-observable, typescript and was migrating from typesafe-actions to redux-toolkit but I was faced with the craziest errors.

This code:
```ts
export type MyEpic = Epic<AnyAction, AnyAction, MyState>;
```
was causing this error elsewhere:
```
Type 'import(".../node_modules/redux-observable/node_modules/rxjs/dist/types/internal/Observable").Observable<any> | undefined' is not assignable to type import(".../node_modules/rxjs/dist/types/internal/Observable").Observable<any> | undefined'.
```

I don't know if there was some yarn issue that led to a broken state, but deleting my yarn.lock and node_modules and doing a clean yarn install fixed it for me.

If anyone else has incomprehensible issues like this, go ahead and try that.

**Comment by MursalovAltun** [comment]
Hey everybody! 

Can we make a creator function for epics creation? Like it's done in NGRX [here](https://github.com/ngrx/platform/blob/a03666b86d970bf375e88239ba302e0fd83fa1e5/modules/effects/src/effect_creator.ts#L54). It would also eliminate ignoreElements or EMPTY for actions that doesn't dispatch anything after the execution.

**Comment by rgbkrk** [comment]
This got me thinking about how maybe the `ofType` operator could check to see if it's passed any `ActionCreatorWithPayload` types so we can infer the type automatically.

**Comment by whazor** [comment]
In my redux-observable-thunk library I have two methods (`filterActions` and `multiMatch`) that are compatible with redux-toolkit in 
[filterActions.ts](https://github.com/Whazor/redux-observable-thunk/blob/2cdfe15aa9c1c90862ddce39f18c2bc9112a7e5e/src/filterActions.ts). It correctly infers types from redux-tookit feel free to try it out (or copy the code).

**Comment by mattborghi** [comment]
I leave this here if someone is looking for an implementation in `v2` of `redux-observable`. It's just a small change in the `middleware` part of `configureStore`.

```javascript
import {
  configureStore,
  Action,
  createSlice,
  PayloadAction,
  combineReducers,
  AnyAction,
} from "@reduxjs/toolkit";
import { createEpicMiddleware, Epic } from "redux-observable";
import { filter, map, delay } from "rxjs/operators";

export const counter = createSlice({
  name: "observable/counter",
  initialState: 0,
  reducers: {
    increment: (state, action: PayloadAction<number>) => state + action.payload,
    decrement: (state, action: PayloadAction<number>) => state - action.payload,
  },
});

const reducer = combineReducers({
  counter: counterReducer,
});

export type MyState = ReturnType<typeof reducers>;
export type MyEpic = Epic<AnyAction, AnyAction, MyState>;

const countEpic: MyEpic = action$ => {
  return action$.pipe(
    filter(counter.actions.increment.match),
    delay(500),
    map((action) => counter.actions.decrement(action.payload))
  );
};

const epicMiddleware = createEpicMiddleware<AnyAction, AnyAction, MyState>();

export const store = configureStore({
  reducer,
  middleware: getDefaultMiddleware => getDefaultMiddleware({thunk: false}).concat(epicMiddleware),
});

epicMiddleware.run(countEpic);
```

**Comment by ratasorin** [comment]
No type errors with this approach ?

**Comment by evertbouw** [comment]
I copied the vite-template-redux project and added redux-observable to it. Check it out here https://github.com/redux-observable/vite-template-redux-observable
